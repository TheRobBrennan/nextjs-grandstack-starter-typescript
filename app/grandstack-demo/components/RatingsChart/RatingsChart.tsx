/* istanbul ignore file */
import { FC } from "react"
import { useQuery } from "@apollo/react-hooks"
import gql from "graphql-tag"

// Components
import Title from "../Title/Title"

// Material UI
import { useTheme } from "@material-ui/core/styles"

// Chart
import {
  Bar,
  XAxis,
  YAxis,
  Label,
  ResponsiveContainer,
  BarChart,
} from "recharts"

// GraphQL
export const GET_DATA_QUERY = gql`
  {
    ratingsCount {
      stars
      count
    }
  }
`

interface IRatingsChart {
  width?: number | string
  height?: number | string
}

const RatingsChart: FC<IRatingsChart> = ({ width, height }) => {
  const theme = useTheme()
  const { loading, error, data } = useQuery(GET_DATA_QUERY)

  if (error) return <p>Error</p>
  if (loading) return <p>Loading</p>

  return (
    <>
      <Title>Ratings Distribution</Title>
      <ResponsiveContainer
        height={height ? height : "100%"}
        width={width ? width : "100%"}
      >
        <BarChart
          data={data.ratingsCount}
          margin={{
            top: 16,
            right: 16,
            bottom: 0,
            left: 24,
          }}
        >
          <XAxis dataKey="stars" stroke={theme.palette.text.secondary} />
          <YAxis stroke={theme.palette.text.secondary}>
            <Label
              angle={270}
              position="left"
              style={{ textAnchor: "middle", fill: theme.palette.text.primary }}
            >
              Count
            </Label>
          </YAxis>
          <Bar dataKey="count" fill={theme.palette.primary.main}></Bar>
        </BarChart>
      </ResponsiveContainer>
    </>
  )
}
export default RatingsChart
