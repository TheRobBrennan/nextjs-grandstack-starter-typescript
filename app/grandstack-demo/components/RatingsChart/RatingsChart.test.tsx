import React from "react"
import { mount } from "enzyme"

import RatingsChart, { GET_DATA_QUERY } from "./RatingsChart"
import { MockedProvider } from "@apollo/client/testing"
import { act } from "@testing-library/react"

describe("GRANDstack RatingsChart component", () => {
  describe("when invoked WITHOUT a specific height or size", () => {
    it("should render the Ratings Distribution chart after receiving data", async () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_DATA_QUERY,
          variables: {},
        },
        result: {
          data: {
            ratingsCount: [
              { stars: 3, count: 2, __typename: "RatingCount" },
              { stars: 4, count: 4, __typename: "RatingCount" },
              { stars: 5, count: 6, __typename: "RatingCount" },
            ],
          },
        },
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      const wrapper = mount(
        <MockedProvider mocks={gqlMocks} addTypename={true}>
          <RatingsChart />
        </MockedProvider>
      )

      // Verify loading state
      expect(wrapper.html()).toContain("Loading")

      // Advance to the next tick in the event loop so our chart can render
      await act(() => {
        return new Promise((resolve) => {
          setTimeout(resolve, 0)
        })
      })
      wrapper.update()

      // Verify our chart has rendered as expected
      expect(wrapper.html()).toMatchSnapshot()
    })
    it("should display an error message if our request resulted in an error", async () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_DATA_QUERY,
          variables: {},
        },
        error: new Error("Uh oh. Something bad happened."),
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      const wrapper = mount(
        <MockedProvider mocks={gqlMocks} addTypename={true}>
          <RatingsChart />
        </MockedProvider>
      )

      // Verify loading state
      expect(wrapper.html()).toContain("Loading")

      // Advance to the next tick in the event loop so our chart can render
      await act(() => {
        return new Promise((resolve) => {
          setTimeout(resolve, 0)
        })
      })
      wrapper.update()

      // Result
      expect(wrapper.html()).toContain("Error")
      expect(wrapper.html()).not.toContain("Ratings Distribution")
    })
  })

  describe("when invoked WITH a specific height and size", () => {
    it("should render the Ratings Distribution chart after receiving data", async () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_DATA_QUERY,
          variables: {},
        },
        result: {
          data: {
            ratingsCount: [
              { stars: 3, count: 2, __typename: "RatingCount" },
              { stars: 4, count: 4, __typename: "RatingCount" },
              { stars: 5, count: 6, __typename: "RatingCount" },
            ],
          },
        },
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      const wrapper = mount(
        <MockedProvider mocks={gqlMocks} addTypename={true}>
          <RatingsChart height={300} width={300} />
        </MockedProvider>
      )

      // Verify loading state
      expect(wrapper.html()).toContain("Loading")

      // Advance to the next tick in the event loop so our chart can render
      await act(() => {
        return new Promise((resolve) => {
          setTimeout(resolve, 0)
        })
      })
      wrapper.update()

      // Verify our chart has rendered as expected
      expect(wrapper.html()).toMatchSnapshot()
    })
    it("should display an error message if our request resulted in an error", async () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_DATA_QUERY,
          variables: {},
        },
        error: new Error("Uh oh. Something bad happened."),
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      const wrapper = mount(
        <MockedProvider mocks={gqlMocks} addTypename={true}>
          <RatingsChart height={300} width={300} />
        </MockedProvider>
      )

      // Verify loading state
      expect(wrapper.html()).toContain("Loading")

      // Advance to the next tick in the event loop so our chart can render
      await act(() => {
        return new Promise((resolve) => {
          setTimeout(resolve, 0)
        })
      })
      wrapper.update()

      // Result
      expect(wrapper.html()).toContain("Error")
      expect(wrapper.html()).not.toContain("Ratings Distribution")
    })
  })
})
