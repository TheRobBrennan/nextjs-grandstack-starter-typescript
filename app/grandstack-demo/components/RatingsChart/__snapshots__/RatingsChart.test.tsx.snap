// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GRANDstack RatingsChart component when invoked WITH a specific height and size should render the Ratings Distribution chart after receiving data 1`] = `"<h2 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom\\">Ratings Distribution</h2><div class=\\"recharts-responsive-container\\" style=\\"width: 300px; height: 300px;\\"><div class=\\"recharts-wrapper\\" style=\\"position: relative; cursor: default; width: 300px; height: 300px;\\"><svg class=\\"recharts-surface\\" width=\\"300\\" height=\\"300\\" viewBox=\\"0 0 300 300\\" version=\\"1.1\\"><defs><clipPath id=\\"recharts3-clip\\"><rect x=\\"84\\" y=\\"16\\" height=\\"254\\" width=\\"200\\"></rect></clipPath></defs><g class=\\"recharts-layer recharts-cartesian-axis recharts-xAxis xAxis\\"><line class=\\"recharts-cartesian-axis-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"200\\" height=\\"30\\" x=\\"84\\" y=\\"270\\" fill=\\"none\\" x1=\\"84\\" y1=\\"270\\" x2=\\"284\\" y2=\\"270\\"></line><g class=\\"recharts-cartesian-axis-ticks\\"><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"200\\" height=\\"30\\" x=\\"84\\" y=\\"270\\" fill=\\"none\\" x1=\\"117.33333333333334\\" y1=\\"276\\" x2=\\"117.33333333333334\\" y2=\\"270\\"></line><text stroke=\\"none\\" width=\\"200\\" height=\\"30\\" x=\\"117.33333333333334\\" y=\\"278\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"middle\\"><tspan x=\\"117.33333333333334\\" dy=\\"0.71em\\">3</tspan></text></g><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"200\\" height=\\"30\\" x=\\"84\\" y=\\"270\\" fill=\\"none\\" x1=\\"184.00000000000003\\" y1=\\"276\\" x2=\\"184.00000000000003\\" y2=\\"270\\"></line><text stroke=\\"none\\" width=\\"200\\" height=\\"30\\" x=\\"184.00000000000003\\" y=\\"278\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"middle\\"><tspan x=\\"184.00000000000003\\" dy=\\"0.71em\\">4</tspan></text></g><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"200\\" height=\\"30\\" x=\\"84\\" y=\\"270\\" fill=\\"none\\" x1=\\"250.66666666666669\\" y1=\\"276\\" x2=\\"250.66666666666669\\" y2=\\"270\\"></line><text stroke=\\"none\\" width=\\"200\\" height=\\"30\\" x=\\"250.66666666666669\\" y=\\"278\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"middle\\"><tspan x=\\"250.66666666666669\\" dy=\\"0.71em\\">5</tspan></text></g></g></g><g class=\\"recharts-layer recharts-cartesian-axis recharts-yAxis yAxis\\"><line class=\\"recharts-cartesian-axis-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"60\\" height=\\"254\\" x=\\"24\\" y=\\"16\\" fill=\\"none\\" x1=\\"84\\" y1=\\"16\\" x2=\\"84\\" y2=\\"270\\"></line><g class=\\"recharts-cartesian-axis-ticks\\"><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"60\\" height=\\"254\\" x=\\"24\\" y=\\"16\\" fill=\\"none\\" x1=\\"78\\" y1=\\"270\\" x2=\\"84\\" y2=\\"270\\"></line><text stroke=\\"none\\" width=\\"60\\" height=\\"254\\" x=\\"76\\" y=\\"270\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"end\\"><tspan x=\\"76\\" dy=\\"0.355em\\">0</tspan></text></g><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"60\\" height=\\"254\\" x=\\"24\\" y=\\"16\\" fill=\\"none\\" x1=\\"78\\" y1=\\"206.5\\" x2=\\"84\\" y2=\\"206.5\\"></line><text stroke=\\"none\\" width=\\"60\\" height=\\"254\\" x=\\"76\\" y=\\"206.5\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"end\\"><tspan x=\\"76\\" dy=\\"0.355em\\">2</tspan></text></g><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"60\\" height=\\"254\\" x=\\"24\\" y=\\"16\\" fill=\\"none\\" x1=\\"78\\" y1=\\"143\\" x2=\\"84\\" y2=\\"143\\"></line><text stroke=\\"none\\" width=\\"60\\" height=\\"254\\" x=\\"76\\" y=\\"143\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"end\\"><tspan x=\\"76\\" dy=\\"0.355em\\">4</tspan></text></g><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"60\\" height=\\"254\\" x=\\"24\\" y=\\"16\\" fill=\\"none\\" x1=\\"78\\" y1=\\"79.5\\" x2=\\"84\\" y2=\\"79.5\\"></line><text stroke=\\"none\\" width=\\"60\\" height=\\"254\\" x=\\"76\\" y=\\"79.5\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"end\\"><tspan x=\\"76\\" dy=\\"0.355em\\">6</tspan></text></g><g class=\\"recharts-layer recharts-cartesian-axis-tick\\"><line class=\\"recharts-cartesian-axis-tick-line\\" stroke=\\"rgba(0, 0, 0, 0.54)\\" width=\\"60\\" height=\\"254\\" x=\\"24\\" y=\\"16\\" fill=\\"none\\" x1=\\"78\\" y1=\\"16\\" x2=\\"84\\" y2=\\"16\\"></line><text stroke=\\"none\\" width=\\"60\\" height=\\"254\\" x=\\"76\\" y=\\"16\\" fill=\\"rgba(0, 0, 0, 0.54)\\" class=\\"recharts-text recharts-cartesian-axis-tick-value\\" text-anchor=\\"end\\"><tspan x=\\"76\\" dy=\\"0.355em\\">8</tspan></text></g></g><text style=\\"text-anchor: middle; fill: rgba(0, 0, 0, 0.87);\\" x=\\"19\\" y=\\"143\\" transform=\\"rotate(270, 19, 143)\\" class=\\"recharts-text recharts-label\\" text-anchor=\\"end\\"><tspan x=\\"19\\" dy=\\"0.355em\\">Count</tspan></text></g><g class=\\"recharts-layer recharts-bar\\"><g class=\\"recharts-layer recharts-bar-rectangles\\"><g class=\\"recharts-layer\\"><g class=\\"recharts-layer recharts-bar-rectangle\\"></g><g class=\\"recharts-layer recharts-bar-rectangle\\"></g><g class=\\"recharts-layer recharts-bar-rectangle\\"></g></g></g></g></svg></div><div style=\\"position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;\\"></div></div>"`;

exports[`GRANDstack RatingsChart component when invoked WITHOUT a specific height or size should render the Ratings Distribution chart after receiving data 1`] = `"<h2 class=\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom\\">Ratings Distribution</h2><div class=\\"recharts-responsive-container\\" style=\\"width: 100%; height: 100%;\\"><div style=\\"position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;\\"></div></div>"`;
